### ISSUE TITLE
Weird behaviour downloading files through PHP code

### ISSUE THREAD

**Issue by gpenin** [nginx/unit#346]
Hi,

Configuration : 
- Debian GNU/Linux : 9.11 (stretch)
- PHP : 7.0.33-0+deb9u6
- NGINX Unit : 1.12.0-1~stretch

Trying to use Owncloud with NGINX Unit, I'm having problems when downloading large files (> 50/100 MB) through PHP code. The file transfer hangs almost every time after having downloaded a few MB, most of the time approximately 10 MB, but sometimes more (50 MB) or sometimes less (2/3 MB). I can reproduce this behavior with a simple PHP code and a random file of 1 GB named test.bin :

```
<?php
$file = '/var/www/test.bin';

if (file_exists($file)) {
    header('Content-Description: File Transfer');
    header('Content-Type: application/octet-stream');
    header('Content-Disposition: attachment; filename="'.basename($file).'"');
    header('Expires: 0');
    header('Cache-Control: must-revalidate');
    header('Pragma: public');
    header('Content-Length: ' . filesize($file));
    readfile($file);
    exit;
}
?>
```

The result :
```
root@server:~# wget http://127.0.0.1:8090/download.php
--2019-11-13 23:15:21--  http://127.0.0.1:8090/download.php
Connexion à 127.0.0.1:8090… connecté.
requête HTTP transmise, en attente de la réponse… 200 OK
Taille : 1073741824 (1,0G) [application/octet-stream]
Sauvegarde en : « download.php »

download.php                                                          1%[=>                                                                                                                                                                 ]  16,34M  --.-KB/s    eta 20m 50s
```

I followed all the following recommendations : 

- NGINX Unit : 
  - "max_body_size": 2147483648

- PHP (recommendations from Owncloud) : 
  - output_buffering = Off
  - memory_limit = 2048M
  - post_max_size = 2048M
  - upload_max_filesize = 2048M

Using a NGINX Unit static file serving configuration, downloading the test.bin file is OK.

**Comment by mar0x** [comment]
Hello,

Please check `unit.log`. Share all errors and warnings if any.

Most likely the issue is caused by unlimited shared memory allocations. I'm currently working on this issue and have a preliminary patch set for testing. I'm appreciate if you try these patches [unit-1.12.0-gh346.tar.gz](https://github.com/nginx/unit/files/3846294/unit-1.12.0-gh346.tar.gz) and let me know the result. You may apply all patches (01-05) for unit-1.12.0 sources. To try latest Unit from GitHub, skip first 2 patches since it is already there.

**Comment by gpenin** [comment]
Hi,

Building unit-1.13.0 sources after applying patches 03, 04 and 05, I can confirm that downloading files through PHP code now works without any problem.

Thanks a lot.

**Comment by mar0x** [comment]
Fixed by df7caf465072e171f88358b9e69c65b76d8efd25 commit.
